base/utf.jl

```diff
@@ -0,0 +1,880 @@
+@param[in]  str::AbstractString
+
+@return     ::UTF16String
+@throws     ArgumentError
+"""
+=#
+function convert(::Type{UTF16String}, str::AbstractString)
+    cnts::UTF_String_Counts = check_string_abs(str)
+    buf = Vector{UInt16}(cnts.cntT+cnts.cnt4+1)
+    out = 0
+    @inbounds for ch in str
+        c = reinterpret(UInt32, ch)
+        if c < 0x10000
+            buf[out += 1] = UInt16(c)
+        elseif c <= 0x10ffff
+            @output_utf16_surr!(buf, out, c)
+        else
+            throw(ArgumentError("invalid Unicode character (0x$(hex(c)) > 0x10ffff)"))
+        end
+    end
+    @inbounds buf[out += 1] = 0 # NULL termination
+    UTF16String(buf)
+end
+
+#=
+"""
+@brief      Converts an AbstractString to UTF-32 encoding
+
+@param[in]  ::Type{UTF32String}
+@param[in]  str::AbstractString
+
+@return     ::UTF32String
+@throws     ArgumentError
+"""
+=#
+function convert(::Type{UTF32String}, str::AbstractString)
+    len = length(str)
+    buf = Vector{Char}(len+1)
+    out = 0
+    @inbounds for ch in str ; buf[out += 1] = ch ; end
+    buf[out += 1] = Char(0) # NULL termination
+    UTF32String(buf)
+end
+
+#=
+@doc """
+@brief      Converts a UTF-8 encoded string to UTF-16 encoding
+
+@param[in]  src::Vector{UInt8}
+
+@return     ::UTF16String
+@throws     ArgumentError
+""" ->
+=#
+function convert(::Type{UTF16String}, src::UTF8String)
+    str = src.data
+    # handle zero length string quickly
+    if sizeof(str) == 0 ; buf = Vector{UInt16}(1) ; buf[1] = 0 ; return UTF16String(buf) ; end
+    # Check that is correct UTF-8 encoding and get number of words needed
+    cnts::UTF_String_Counts = check_string_utf8(str)
+    len = cnts.cntT + cnts.cnt4
+    buf = Vector{UInt16}(len+1)
+    buf[len+1] = 0
+    # Optimize case where no characters > 0x7f
+    cnts.flags == 0 && return UTF16String(copy!(buf, str))
+    out::UInt = 0
+    pos::UInt = 0
+    @inbounds while out < len
+        ch::UInt32 = str[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle range 0x80-0x7ff
+        elseif ch < 0xe0
+            buf[out += 1] = @get_utf8_2!(str, pos, ch)
+        # Handle range 0x800-0xffff
+        elseif ch < 0xf0
+            buf[out += 1] = @get_utf8_3!(str, pos, ch)
+        # Handle range 0x10000-0x10ffff
+        else
+            ch = @get_utf8_4!(str, pos, ch)
+            @output_utf16_surr!(buf, out, ch)
+        end
+    end
+    UTF16String(buf)
+end
+
+#=
+@doc """
+@brief      Reencodes a UTF-16 or UTF-32 encoded string using UTF-8 encoding
+
+@param[in]  str::Union(Vector{UInt16}, Vector{UInt32})
+
+@return     ::UTF8String
+@throws     ArgumentError
+""" ->
+=#
+function convert(::Type{UTF8String}, str::Vector{UInt16})
+    len = sizeof(str)
+    # handle zero length string quickly
+    len == 0 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf16(str, len>>>1)
```

Comment:
I would find it clearer to write `len>>>1` as `len/2` (same elsewhere), unless you can measure a real performance difference. Compared to the work you have to do, this kind of optimization sounds premature to me.